Userland/Libraries/LibWeb/Layout/TextNode.cpp

```diff
@@ -103,35 +103,61 @@ void TextNode::paint_cursor_if_needed(PaintContext& context, const LineBoxFragme
     context.painter().draw_rect(cursor_rect, computed_values().color());
 }
 
+// NOTE: This collapes whitespace into a single ASCII space if collapse is true. If previous_is_empty_or_ends_in_whitespace, it also strips leading whitespace.
 void TextNode::compute_text_for_rendering(bool collapse, bool previous_is_empty_or_ends_in_whitespace)
 {
-    if (!collapse) {
-        m_text_for_rendering = dom_node().data();
+    auto& data = dom_node().data();
+    if (!collapse || data.is_empty()) {
+        m_text_for_rendering = data;
         return;
     }
 
-    // Collapse whitespace into single spaces
-    auto utf8_view = Utf8View(dom_node().data());
-    StringBuilder builder(dom_node().data().length());
-    auto it = utf8_view.begin();
-    auto skip_over_whitespace = [&] {
-        auto prev = it;
-        while (it != utf8_view.end() && is_ascii_space(*it)) {
-            prev = it;
-            ++it;
+    // NOTE: A couple fast returns to avoid unnecessarily allocating a StringBuilder.
+    if (data.length() == 1) {
+        if (is_ascii_space(data[0])) {
+            if (previous_is_empty_or_ends_in_whitespace)
+                m_text_for_rendering = String::empty();
+            else
+                m_text_for_rendering = " ";
```

Comment:
I suggest keeping a `static String` with this exact content on hand. Then we don't need to keep creating new `String` objects here at all :)